### PR TITLE
Add validation to topology updates and server updates/deletions

### LINKED ISSUE #5171: Need to validate PUT/DELETE /servers satisfying a Topology-based Delivery Service's Required Capabilities
## I'm submitting a ...

-  new feature / enhancement request

## Traffic Control components affected ...

-  Traffic Ops

## Current behavior:

On PUT /servers, it is possible to change the CDN of a server to unsatisfy a Topology-based Delivery Service's Required Capabilities.

## New behavior:

We need validation to prevent that.

## Minimal reproduction of the problem with instructions:

Scenario 1:
1. Have a Topology-based Delivery Service
2. Create a second CDN
3. Move the last Server in a Cache Group used by that Top-based DS to CDN 2.

Scenario 2:
1. Have a Topology-based Delivery Service
3. Delete the last Server in a Cache Group used by that Top-based DS

Validation should prevent both scenarios.

Discovered by @dpham692

### LINKED ISSUE #5286: Server update Topology safeties can be bypassed at lower API versions
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
When modifying a server's Cache Group assignment, the API only checks if this invalidates any Topologies if the client made the request at API version 3.

## Expected behavior:
This restriction should be in place unilaterally across API versions.

## Minimal reproduction of the problem with instructions:
In CDN-in-a-Box, create a new EDGE Cache Group and move the `edge` server into it; notice this operation fails in API v3, but not if you request API v2 or lower.

### PR DESCRIPTION
**Pull request by rawlinp**
## What does this PR (Pull Request) do?
Adds the following validation to topology updates and server updates/deletions:

Topologies must contain at least 1 server per cachegroup in each of the
CDNs of any assigned delivery services. Otherwise, a delivery service's
topology is not truly representative of reality.

- [x] This PR fixes #5171 and fixes #5286 

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?
Review the added tests, ensure they pass.

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] validation, no docs necessary
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Review by ocket8888 — APPROVED** [review]
This all looks solid to me.

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 1]
```diff
@@ -881,6 +881,51 @@ WHERE
...
+	res := make([]int, len(cdnIDs))
+	for i, id := range cdnIDs {
+		res[i] = int(id)
+	}
+	return res, nil
+}
+
+// CheckCachegroupHasTopologyBasedDeliveryServicesOnCDN returns true if the given cachegroup is assigned to
+// any topologies with delivery services assigned on the given CDN.
+func CachegroupHasTopologyBasedDeliveryServicesOnCDN(tx *sql.Tx, cachegroupID int, CDNID int) (bool, error) {
```
Comment (zrhoffman) [inline traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:908]: Godoc doesn't match function name

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 2]
```diff
@@ -881,6 +881,51 @@ WHERE
...
+	q := `
+SELECT EXISTS(
+  SELECT
+    1
```
Comment (zrhoffman) [inline traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:912]: The `1` is unnecessary, can be just `SELECT FROM`

**Review by zrhoffman — COMMENTED** [review]
I'm running into a bug where, if a Topology's Cache Group is already empty with respect to the CDN used by a Delivery Service that uses that Topology, you cannot move servers into that Cache Group.

Steps to reproduce in CDN in a Box:

1. Clone `ATS_MID_TIER_CACHE` to a new profile for the `ALL` CDN

2. In the database, make server `mid-01` be in CDN `ALL`
    ```sql
    UPDATE "server" s
    SET cdn_id = (SELECT id FROM cdn WHERE cdn."name" = 'ALL')
    WHERE s.host_name = 'mid-01'
    ```

3. Go to Servers -> `mid-01` and change the CDN to `CDN-in-a-Box`, click `Update`

4. Receive error:
    ```
    server is the last one in its cachegroup, which is used by a topology, so it cannot be moved to another cachegroup: cachegroups with no servers in them: CDN_in_a_Box_Mid-01 (in topologies: demo1-top)
    ```

However, if the `CDN_in_a_Box_Mid-01` Cache Group is not empty with respect to the `CDN-in-a-Box` CDN before attempting the update, the update will succeed.

**Pull request merged**